src/commands/key2ds.rs

```diff
@@ -0,0 +1,166 @@
+use std::{fs::File, io::Write, path::PathBuf};
+
+use clap::{builder::ValueParser, Parser};
+use domain::{
+    base::{
+        iana::{DigestAlg, SecAlg},
+        zonefile_fmt::ZonefileFmt,
+        Record,
+    },
+    rdata::Ds,
+    validate::DnskeyExt,
+    zonefile::inplace::{Entry, ScannedRecordData},
+};
+
+use crate::error::Error;
+
+#[derive(Clone, Debug, Parser)]
+#[command(version)]
+pub struct Key2ds {
+    /// ignore SEP flag (i.e. make DS records for any key)
+    #[arg(short = 'f')]
+    ignore_sep: bool,
+
+    /// do not write DS records to file(s) but to stdout
+    #[arg(short = 'n')]
+    write_to_stdout: bool,
+
+    /// use SHA-1 for the DS hash
+    #[arg(short = '1', overrides_with_all = ["one", "two", "four", "algorithm"])]
+    one: bool,
+
+    /// use SHA-256 for the DS hash
+    #[arg(short = '2', overrides_with_all = ["one", "two", "four", "algorithm"])]
+    two: bool,
+
+    /// use SHA-384 for the DS hash
+    #[arg(short = '4', overrides_with_all = ["one", "two", "four", "algorithm"])]
+    four: bool,
+
+    /// algorithm to use for digest
+    #[arg(
+        short = 'a',
+        long = "algorithm",
+        overrides_with_all = ["one", "two", "four", "algorithm"],
+        value_parser = ValueParser::new(parse_digest_alg)
+    )]
+    algorithm: Option<DigestAlg>,
+
+    /// Keyfile to read
+    #[arg()]
+    keyfile: PathBuf,
+}
+
+pub fn parse_digest_alg(arg: &str) -> Result<DigestAlg, Error> {
+    if let Ok(num) = arg.parse() {
+        let alg = DigestAlg::from_int(num);
+        if alg.to_mnemonic().is_some() {
+            Ok(alg)
+        } else {
+            Err(Error::from("unknown algorithm number"))
+        }
+    } else {
+        DigestAlg::from_mnemonic(arg.as_bytes()).ok_or(Error::from("unknown algorithm mnemonic"))
+    }
+}
+
+impl Key2ds {
+    pub fn execute(self) -> Result<(), Error> {
+        let mut file = std::fs::File::open(&self.keyfile).map_err(|e| {
+            format!(
+                "Failed to open public key file \"{}\": {e}",
+                self.keyfile.display()
+            )
+        })?;
+        let zonefile = domain::zonefile::inplace::Zonefile::load(&mut file).unwrap();
+        for entry in zonefile {
+            let entry = entry.map_err(|e| {
+                format!(
+                    "Error while reading public key file \"{}\": {e}",
```

Comment:
This seems a strange error to me. There was a problem iterating over a zone file but the error is about reading a public key file?

Reply:
key2ds is usually used on key files containing only a single dnskey (e.g. generated by keygen). These files are in zonefile format, and you could use a single zone file containing all dnskeys you want to get DS records from. So maybe saying "Error while reading public key *from* file" will keep it ambiguous yet clear enough to apply to both reading from an actual zone file (used as a normal zone file), and reading from a keyfile (that is in zone file format)?

Reply:
Ah!

Reply:
I like @mozzieongit's suggestion. Will push a fix for that.